File: Libraries/Megamark Python 2.7/Megamark.py
'''
The MMM.py script contains all bindings to control the Megamark robot with Python 2.7.
Include this script in same directory of your other Python scripts referencing the Megamark.
Note: The script Serial.ino from the Megamark Library must be loaded on the Arduino to work!
'''

### Load Libraries ###
import serial    #Needed to communicate with the robot's Arduino Mega 2560.
                 #Install it here: https://pypi.python.org/pypi/pyserial/2.7
				 
import threading #Used for concurrent robot communication with other processes. 
                 #More info: https://docs.python.org/2/library/threading.html
       
import time      #Used to create delays for proper timing of robot actions.
                 #More info: https://docs.python.org/2/library/time.html

import sys       #Used for certain specific calls to the system.
                 #More info: https://docs.python.org/2/library/sys.html

### The Multipurpose Mobile Manipulator Class ###
class Megamark:
  # Incoming Data Parsing Variables
  dataStr = ""
  dataIndex = 0
  data = [ 0,0 ] #[ Left Laser, Right Laser ]
  
  # Resets all values
  def reset(self):
    #Wheels (-255 to 255)
    self.leftWheel = 0
    self.rightWheel = 0
    #Shoulders (0 to 180)
    self.leftShoulder = 90
    self.rightShoulder = 90
    #Elbows (0 to 180)
    self.leftElbow = 90
    self.rightElbow =90
    #Left Grippers (0 to 180)
    self.L1 = 90
    self.L2 = 90
    self.L3 = 90
    self.L4 = 90
    self.L5 = 90
    self.L6 = 90
    #Right Grippers (0 to 180)
    self.R1 = 90
    self.R2 = 90
    self.R3 = 90
    self.R4 = 90
    self.R5 = 90
    self.R6 = 90
    #Lasers (variable based on sensor, usually 0-1500 mm)
    self.leftLaser = -1;
    self.rightLaser = -1;
    
  # helper function clamps a value to a min/max
  def clamp(self, value, minimum, maximum):
    return min(max(value, minimum), maximum)
  # helper function clamps all robot values
  def clampAll(self):
    # Clamp Wheels
    self.leftWheel = self.clamp(self.leftWheel,-255,255)
    self.rightWheel = self.clamp(self.rightWheel,-255,255)
    # Clamp Shoulders
    self.leftShoulder = self.clamp(self.leftShoulder,0,180)
    self.rightShoulder = self.clamp(self.rightShoulder,0,180)
    # Clamp Elbows
    self.leftElbow = self.clamp(self.leftElbow,0,180)
    self.rightElbow = self.clamp(self.rightElbow,0,180)
    # Clamp Left Grippers 
    self.L1 = self.clamp(self.L1,0,180)
    self.L2 = self.clamp(self.L2,0,180)
    self.L3 = self.clamp(self.L3,0,180)
    self.L4 = self.clamp(self.L4,0,180)
    self.L5 = self.clamp(self.L5,0,180)
    self.L6 = self.clamp(self.L6,0,180)
    # Clamp Right Grippers 
    self.R1 = self.clamp(self.R1,0,180)
    self.R2 = self.clamp(self.R2,0,180)
    self.R3 = self.clamp(self.R3,0,180)
    self.R4 = self.clamp(self.R4,0,180)
    self.R5 = self.clamp(self.R5,0,180)
    self.R6 = self.clamp(self.R6,0,180)
  # helper function maps one range of values to another
  def translate(self, value, inMin, inMax, outMin, outMax):
    # Clamp value within input range
    value = self.clamp(value, inMin, inMax)
    # Figure out how 'wide' each range is
    inSpan = inMax - inMin
    outSpan = outMax - outMin
    # Convert the left range into a 0-1 range (float)
    valueScaled = float(value - inMin) / float(inSpan)
    # Convert the 0-1 range into a value in the right range.
    return outMin + (valueScaled * outSpan) 
    
  # spin motors in meters/sec (-0.2 to 0.2 m/s)
  def setWheelVelocities(self, leftSpeed, rightSpeed):
    #convert speed to robot units
    self.leftWheel = int(self.translate(leftSpeed, -0.2,0.2, -255.0,255.0))
    self.rightWheel = int(self.translate(rightSpeed, -0.2,0.2, -255.0,255.0))   
  # rotate left shoulder in degrees (0 to 120 degrees)
  def rotateLeftShoulder(self, angle):
    self.leftShoulder = int(self.translate(angle, 0,120, 0,180))
  # rotate right shoulder in degrees (0 to 120 degrees)
  def rotateRightShoulder(self, angle):
    self.rightShoulder = int(self.translate(angle, 0,120, 0,180)) 
  # rotate left elbow in degrees (-60 to 60 degrees)
  def rotateLeftElbow(self, angle):
    self.leftElbow = int(self.translate(angle, -60,60, 0,180))
  # rotate right elbow in degrees (-60 to 60 degrees)
  def rotateRightElbow(self, angle):
    self.rightElbow = int(self.translate(angle, -60,60, 0,180)) 
  # control left grippers (Up to 6 servos, all range from 0-180)
  def setLeftGrippers(self, l1 = 0,l2 = 0,l3 = 0,l4 = 0,l5 = 0,l6 = 0):
    self.L1 = l1;
    self.L2 = l2;
    self.L3 = l3;
    self.L4 = l4;
    self.L5 = l5;
    self.L6 = l6;
  # control right grippers (Up to 6 servos, all range from 0-180)
  def setRightGrippers(self,r1 = 0,r2 = 0,r3 = 0,r4 = 0,r5 = 0,r6 = 0):
    self.R1 = r1;
    self.R2 = r2;
    self.R3 = r3;
    self.R4 = r4;
    self.R5 = r5;
    self.R6 = r6;
  # get left laser sensor data (variable based on sensor, usually 0-1500 mm)
  def getLeftLaser(self):
    self.leftLaser = self.data[0]
    return self.leftLaser
  # get right laser sensor data (variable based on sensor, usually 0-1500 mm)
  def getRightLaser(self):  
    self.rightLaser = self.data[1]
    return self.rightLaser
  
  # Reads data coming from Arduino
  def read(self):
    #Make backup if data in case parsing fails.
    oldData = list(self.data)
    
    #Read data
    try:
      for chr in self.ser.readline():   
        #begin parsing
        if(chr == '{'):
          self.dataStr = ""
          self.dataIndex = 0 
        #extract number from data
        elif(chr == ',' or chr == '}'):
          self.data[self.dataIndex] = int(self.dataStr);
          self.dataIndex+=1; self.dataStr = "";
          #Finished reading message
          if(chr == '}'): 
            self.dataStr = "" 
            self.dataIndex = 0
        #add char to current string
        elif (chr.isspace() == False and chr != '\n'):
          self.dataStr += chr;
    except:
      self.data = oldData
      self.ser.flushInput();
    
  # Sends data to Arduino in Robot units, which updates position 
  def write(self):
    # Clamp all values before sending data.
    self.clampAll();  
	  
    # Prepare data
             #Wheels
    data = [ self.leftWheel, self.rightWheel,
             #Shoulders
             self.leftShoulder, self.rightShoulder, 
             #Elbows
             self.leftElbow, self.rightElbow,
             #Left Grippers
             self.L1, self.L2, self.L3, self.L4, self.L5, self.L6,
             #Right Grippers
             self.R1, self.R2, self.R3, self.R4, self.R5, self.R6 ]
                             
    # begin dataString
    dataString = "{ "
    # Add values to data list
    for dataIndex in range(len(data)):
      # Add data as a string
      dataString += str(int(data[dataIndex]))
      if (dataIndex < len(data)-1):
        dataString += ", "; # to next value
      else: 
        dataString += " }"  # finish string
      
    # send data to Arduino
    try:
      self.ser.flush()
      self.ser.flushInput()
      self.ser.write(dataString)
    except:
      print("Could not update robot at " + self.ser.name + "! Exiting.")
      print("(Check USB connection to Arduino Mega and restart program.)")
      self.ser.close()
      quit()
  
  # Writer function that runs continuously on another thread.
  def writer(self):
    while True:
      self.write();
      time.sleep(0.1);
  # Reader function that runs continuously on another thread.
  def reader(self):
    while True:
      self.read(); 
      time.sleep(0.1);
  
  # Initializes the MMM class  
  def __init__(self, portName):
    #Create a new serial connection to the robot's Arduino Mega at the specified port:
    try:
      self.ser = serial.Serial(portName,57600,timeout=0)
    except:
      print("Could not connect to robot at " + portName + "! Exiting.")
      print("(Check USB connection to Arduino Mega and restart program.)")
      quit()

    print("Successfully connected to robot at port " + portName + ".")	  
    time.sleep(2)       #Wait two seconds to allow connection to initialize.    
    self.ser.readline() #Clear the serial buffer by reading one line.
    self.reset()        #Initialize all robot positions.
    
    #Create a new serial writer thread to update robot actuators continuously:
    try: 
      #begin threads to update continuously:
      self.writeThread = threading.Thread(target=self.writer, args=())
      self.writeThread.daemon = True            
      self.writeThread.start()
    except:
      print("Could not begin writer thread for serial communication! Exiting.")
      self.ser.close()
      quit()
	
    #Create a new serial reader thread to get robot data continuously:
    try: 
      self.readThread = threading.Thread(target=self.reader, args=())
      self.readThread.daemon = True            
      self.readThread.start()	
    except:
      print("Could not begin reader thread for serial communication! Exiting.")
      self.ser.close()
      quit()

  # Function to safely terminate the Megamark instance.
  def exit(self):
    print("Megamark instance is now attempting to exit...")
    try:
      print("Closing serial connection to Arduino...")
      self.ser.close()
      print("Megamark instance successfully closed.")
    except:
      print("Megamark failed to quit. Please terminate the program manually.")
      
      



  
		

File: Libraries/Megamark Unity/Assets/Megamark Controller/Core/Megamark.cs
using CielaSpike;
using UnityEngine;
using UnityEngine.UI;
using System;
using System.IO.Ports;
using System.Collections;

public class Megamark : MonoBehaviour {

	//Asynchronous threading variables
	public float initializationTime = 2; //Time before data is serialized and sent. Set inside inspector.
	public int refreshDelay = 5; //Time between each data sending in miliseconds
	private Task writerThread;
	private Task readerThread; 

	//Arduino variables
	private SerialPort serial;
	public string portName;
	public int timeout = 1000;
	public int baudrate = 57600;
	public bool connected = false;
	public bool connectOnAwake = false;

	// Incoming Data Parsing Variables
	string dataStr = "";
	private int dataIndex = 0;
	public int[] data = { 0, 0 }; // [ Left Laser, Right Laser ]

	// Wheels (-255 to 255)
	public int leftWheel, rightWheel = 0;
	// Shoulders (0 to 180)
	public int leftShoulder, rightShoulder = 90;
	// Elbows (0 to 180)
	public int leftElbow, rightElbow = 90;
	// Left Grippers (0 to 180)
	public int L0, L1, L2, L3, L4, L5 = 90;
	// Right Grippers (0 to 180)
	public int R0, R1, R2, R3, R4, R5 = 90;
	// Lasers (variable based on sensor, usually 0-1500 mm)
	public int leftLaser, rightLaser = -1;
		
	// Sets all servo values to their initial positions: both arms parallel to the ground to the left and right of the robot
	public void reset() {
		//Wheels (-255 to 255)
		leftWheel = 0;
		rightWheel = 0;
		//Shoulders (0 to 180)
		leftShoulder = 90; 
		rightShoulder = 90;
		//Elbows (0 to 180)
		leftElbow = 90; 
		rightElbow = 90;
		//Left Grippers (0 to 180)
		L0 = 90; L1 = 90; L2 = 90; L3 = 90; L4 = 90; L5 = 90;
		//Right Grippers (0-180)
		R0 = 90; R1 = 90; R2 = 90; R3 = 90; R4 = 90; L5 = 90;
		// Lasers (variable based on sensor, usually 0-1500 mm)
		leftLaser = -1;
		rightLaser = -1;
	}
	// Helper function that clamps all robot values
	public void ClampAll() {
		// Clamp Wheels
		leftWheel = Mathf.Clamp(leftWheel,-255,255);
		rightWheel = Mathf.Clamp(rightWheel,-255,255);
		// Clamp Shoulders
		leftShoulder = Mathf.Clamp(leftShoulder,0,180);
		rightShoulder = Mathf.Clamp(rightShoulder,0,180);
		// Clamp Elbows
		leftElbow = Mathf.Clamp(leftElbow,0,180);
		rightElbow = Mathf.Clamp(rightElbow,0,180);
		// Clamp Left Grippers 
		L0 = Mathf.Clamp(L0,0,180);
		L1 = Mathf.Clamp(L1,0,180);
		L2 = Mathf.Clamp(L2,0,180);
		L3 = Mathf.Clamp(L3,0,180);
		L4 = Mathf.Clamp(L4,0,180);
		L5 = Mathf.Clamp(L5,0,180);
		// Clamp Right Grippers 
		R0 = Mathf.Clamp(R0,0,180);
		R1 = Mathf.Clamp(R1,0,180);
		R2 = Mathf.Clamp(R2,0,180);
		R3 = Mathf.Clamp(R3,0,180);
		R4 = Mathf.Clamp(R4,0,180);
		R5 = Mathf.Clamp(R5,0,180);
	}
	// Helper function maps one range of values to another
	int Translate(float value, float inMin, float inMax, float outMin, float outMax) {
		// Clamp value within input range
		value = Mathf.Clamp(value, inMin, inMax);
		// Figure out how 'wide' each range is
		float inSpan = inMax - inMin;
		float outSpan = outMax - outMin;
		// Convert the left range into a 0-1 range (float)
		float valueScaled = (value - inMin) / inSpan;
		// Convert the 0-1 range into a value in the right range.
		return (int)(outMin + (valueScaled * outSpan)); 
	}

	// spin motors in meters/sec (-0.2 to 0.2 m/s)
	public void setWheelVelocities(float leftSpeed, float rightSpeed) {
		//convert speed to robot units
		leftWheel = Translate(leftSpeed, -0.2f, 0.2f, -255, 255);
		rightWheel = Translate(rightSpeed, -0.2f, 0.2f, -255, 255);
	}
		
	// rotate left shoulder in degrees (0 to 120 degrees)
	public void rotateLeftShoulder(float angle) {
		leftShoulder = Translate(angle, 0,120, 0,180);
	}
	// get left shoulder in degrees (0 to 120 degrees)
	public  float getLeftShoulder() {
		return Translate(leftShoulder, 0,180, 0,120);
	}

	// rotate right shoulder in degrees (0 to 120 degrees)
	public void rotateRightShoulder(float angle) {
		rightShoulder = Translate(angle, 0,120, 0,180);
	}
	// get right shoulder in degrees (0 to 120 degrees)
	public float getRightShoulder() {
		return Translate(rightShoulder, 0,180, 0,120);
	}

	// rotate left elbow in degrees (-60 to 60 degrees)
	public void rotateLeftElbow(float angle) {
		leftElbow = Translate(angle, -60,60, 0,180);
	}
	// get left elbow in degrees (-60 to 60 degrees)
	public float getLeftElbow() {
		return Translate(leftElbow, 0,180, -60,60);
	}

	// rotate right elbow in degrees (-60 to 60 degrees)
	public void rotateRightElbow(float angle) {
		rightElbow = Translate(angle, -60,60, 0,180);
	}
	// get right elbow in degrees (-60 to 60 degrees)
	public float getRightElbow() {
		return Translate(rightElbow, 0,180, -60,60);
	}

	// control left grippers (Up to 6 servos, all range from 0-180)
	public void setLeftGrippers(int l0 = 0,int l1 = 0,int l2 = 0,int l3 = 0,int l4 = 0, int l5 = 0) {
		L0 = l0; L1 = l1; L2 = l2; L3 = l3; L4 = l4; L5 = l5;
	}
	// control right grippers (Up to 6 servos, all range from 0-180)
	public void setRightGrippers(int r0 = 0,int r1 = 0,int r2 = 0,int r3 = 0,int r4 = 0, int r5 = 0) {
		R0 = r0; R1 = r1; R2 = r2; R3 = r3; R4 = r4; R5 = r5;
	}
	// get left laser sensor data (variable based on sensor, usually 0-1500 mm)
	public int getLeftLaser() {
		leftLaser = data[0];
		return leftLaser;
	}
	// get right laser sensor data (variable based on sensor, usually 0-1500 mm)
	public int getRightLaser() {
		rightLaser = data[1];
		return rightLaser;
	}	

	// Sends data to Arduino in Robot units, which updates position 
	public void write() {
		// Clamp all values before sending data.
		ClampAll();

		// Prepare data to send:
		//Wheels
		int[] send = { leftWheel, rightWheel, 
			//Shoulders
			leftShoulder, rightShoulder,
			//Elbows
			leftElbow, rightElbow,
			//Left Grippers
			L0, L1, L2, L3, L4, L5,
			//Right Grippers
			R0, R1, R2, R3, R4, R5 };

		// begin dataString
		string dataString = "{ ";
		// Add values to data list
		for (int i = 0; i < send.Length; i++) {
			// Add data as a string
			dataString += send[i].ToString();
			if (i < send.Length - 1) {
				dataString += ", "; // to next value
			} else {
				dataString += " }"; // finish string
			}
		}
		// send data to Arduino
		try {
			serial.Write(dataString);
		} catch (Exception e) {
			if (serial != null && serial.IsOpen) {
				//Debug.LogWarning (e);
				serial.DiscardInBuffer ();
				serial.DiscardOutBuffer ();
			}
		}
	}
	// Sends data to Arduino in Robot units, which updates position 
	public void read() {
		// Make backup if data in case parsing fails:
		int[] oldData = data;    
		// Read data
		try {
			String str = "";
			str = serial.ReadLine();
			for (int i = 0; i < str.Length; i++) {
				// get ith character of serial read string
				char chr = str[i];
				// begin parsing
				if(chr == '{') {
					dataStr = "";
					dataIndex = 0; 
				} 
				// extract number from data
				else if(chr == ',' || chr == '}') {
					data[dataIndex] = int.Parse(dataStr);
					dataIndex += 1; dataStr = "";
					// Finished reading message
					if(chr == '}') { 
						dataStr = "";
						dataIndex = 0;
					}
				}
				// add char to current string
				else if (!char.IsWhiteSpace(chr) && chr != '\n') {
					dataStr += chr;
				}
			}
		} catch(Exception e) {
			data = oldData;
			if (serial != null && serial.IsOpen) {
				//Debug.LogWarning (e);
				serial.DiscardInBuffer ();
				serial.DiscardOutBuffer ();
			}
		}
	}

	// A threaded loop that updates the Arduino through serial connection
	IEnumerator Writer(){
		//continuously update the serial writer thread.
		while (serial.IsOpen) {
			write(); 
			yield return new WaitForSeconds((float)refreshDelay / 1000);
		}
	}
	// A threaded loop that updates the Arduino through serial connection
	IEnumerator Reader(){
		//continuously update the serial reader thread.
		while (serial.IsOpen) {
			read();
			yield return new WaitForSeconds((float)refreshDelay / 1000);
		}
	}
	// Starts reader/writer threads.
	private void startThreads() {
		this.StartCoroutineAsync(Writer(), out writerThread);
		this.StartCoroutineAsync(Reader(), out readerThread);
	}

	// Attempt connection to Arduino through the specified port
	public void Connect () {
		if (!connected) {
			//Create a new Serial port
			serial = new SerialPort (portName, baudrate, Parity.None, 8, StopBits.One);
			serial.WriteTimeout = timeout;
			serial.ReadTimeout = timeout;
			try {
				Debug.Log ("Attempting to connect to robot at port " + portName + "...");
				serial.Open ();
				Debug.Log ("Successfully connected to robot at port " + portName + ".");
				connected = true;
				Invoke ("startThreads", initializationTime);
			} catch (System.Exception e) {
				Debug.LogException (e);
				Debug.Log ("Could not connect to robot at " + portName + "! Exiting.");
				Debug.Log ("(Check USB connection to Arduino Mega and restart program.)");
			}
			reset ();
		} else {
			Debug.Log ("Robot is already connected at " + portName + ".");
			Debug.Log ("(Use another port and/or another instance of the Megamark script.)");
		}
	}

	// Close the serial connection and destroy all serial threads
	public void Disconnect () {
		if (serial.IsOpen) {
			writerThread.Cancel();
			readerThread.Cancel();
			serial.Close();
			connected = false;
			Debug.Log("Disconnected.\n");
		}
	}
	// Making sure to disconnect threads on exit.
	public void OnApplicationQuit() { Disconnect(); }

	// If connectOnAwake is enabled, connect to the robot upon playing the program
	public void Start() { if (connectOnAwake) { Connect(); } }

}
